lib/pbench/server/cache_manager.py

```diff
@@ -82,6 +95,59 @@ def __str__(self) -> str:
         return f"An error occurred while changing file permissions of {self.tarball}: {self.error}"
 
 
+class CacheType(Enum):
+
+    FILE = auto()
+    DIRECTORY = auto()
+    SYMLINK = auto()
+    OTHER = auto()
+
+
+@dataclass
+class CacheObject:
+
+    name: str = None
+    location: Path = None
+    resolve_path: Optional[Path] = None
+    resolve_type: str = None
+    size: int = None
+    type: str = CacheType
+    dir_path: InitVar[Path] = None
+    path: InitVar[Path] = None
+
+    def __post_init__(self, dir_path, path):
+        self.name = path.name
+        self.location = path.relative_to(dir_path)
+        if path.is_symlink():
+            self.type = CacheType.SYMLINK
+            try:
+                resolve_path = path.resolve(strict=True)
+            except FileNotFoundError:
+                resolve_path = path.readlink()
+
+            try:
+                self.resolve_path = resolve_path.relative_to(dir_path)
+            except ValueError:
+                self.resolve_path = resolve_path
+                self.resolve_type = CacheType.SYMLINK
+            else:
+                if not resolve_path.exists():
+                    self.resolve_type = CacheType.SYMLINK
+                elif resolve_path.is_dir():
+                    self.resolve_type = CacheType.DIRECTORY
+                elif resolve_path.is_file():
+                    self.resolve_type = CacheType.FILE
+                else:
+                    self.resolve_type = CacheType.OTHER
+        elif path.is_file():
+            self.type = CacheType.FILE
+            self.size = path.stat().st_size
+        elif path.is_dir():
+            self.type = CacheType.DIRECTORY
+        else:
+            self.type = CacheType.OTHER
```

Comment:
So, let me take one more stab at this.  (Sorry for the rough ride, but we're learning as we go.... 🥴)

```python
def make_cache_info(dir_path: Path, path: Path) -> CacheObject:
    relative_path: Optional[Path] = None
    resolve_type: Optional[CacheType] = None
    size: Optional[int] = None

    if path.is_symlink():
        ftype = CacheType.SYMLINK
	try:
            link_path = path.readlink()
            if link_path.is_absolute():
	        raise ValueError("symlink path is absolute")
            resolve_path = path.resolve(strict=True)
            relative_path = resolve_path.relative_to(dir_path)
	except (FileNotFound, ValueError):
            relative_path = link_path
            resolve_type = CacheType.SYMLINK
	else:
            if resolve_path.is_dir():
                resolve_type = CacheType.DIRECTORY
	    elif resolve_path.is_file():
                resolve_type = CacheType.FILE
            else:
                resolve_type = CacheType.OTHER
    elif path.is_file():
        ftype = CacheType.FILE
        size = path.stat().st_size
    elif path.is_dir():
        ftype = CacheType.DIRECTORY
    else:
        ftype = CacheType.OTHER

    return CacheObject(
        name=path.name,
        location=path.relative_to(dir_path),
        resolve_path=relative_path,
        resolve_type=resolve_type,
        size=size,
        type=ftype,
    )

```
That is, if the `path` is a symlink, then we read the contents of the link; if the link is an absolute reference or the link refers to a file which doesn't exist or it refers to an entry outside the `dir_path` tree, then we set the "relative path" to what we read from the symlink file and we mark the resolution as a `SYMLINK`; otherwise, we test the resolved path and mark it as a `DIRECTORY`, `FILE`, or `OTHER` as appropriate.  If the `path` is a regular file, a directory, or something else, we handle it.  Then, finally, we take the `name, `location`, `relative_path`, `resolve_type`, `size`, and `ftype` and roll them up into a `CacheObject` which we then return.